Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp

```diff
@@ -278,49 +280,82 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
 #if ENABLE(TRACKING_PREVENTION)
-    if (shouldBlockCookies(webFrame.get(), document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess)) {
-        completionHandler({ });
-        return;
+    if (frame->loader().client().isRemoteWorkerFrameLoaderClient()) {
+        if (shouldBlockCookies(nullptr, document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess)) {
+            completionHandler({ });
+            return;
+        }
+    } else {
+        RefPtr webFrame = document.frame() ? WebFrame::fromCoreFrame(*document.frame()) : nullptr;
+        if (!webFrame || !webFrame->page()) {
+            completionHandler({ });
+            return;
+        }
+
+        if (shouldBlockCookies(webFrame.get(), document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess)) {
+            completionHandler({ });
+            return;
+        }
+
+        frameID = webFrame->frameID();
+        pageID = webFrame->page()->identifier();
     }
 #endif
 
     auto sameSiteInfo = CookieJar::sameSiteInfo(document, IsForDOMCookieAccess::Yes);
     auto includeSecureCookies = CookieJar::shouldIncludeSecureCookies(document, url);
-    auto frameID = webFrame->frameID();
-    auto pageID = webFrame->page()->identifier();
 
-    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::CookiesForDOMAsync(document.firstPartyForCookies(), sameSiteInfo, url, frameID, pageID, includeSecureCookies, applyTrackingPreventionInNetworkProcess, shouldRelaxThirdPartyCookieBlocking(webFrame.get()), options), WTFMove(completionHandler));
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::CookiesForDOMAsync(document.firstPartyForCookies(), sameSiteInfo, url, frameID, pageID, includeSecureCookies, applyTrackingPreventionInNetworkProcess, shouldRelaxThirdPartyCookieBlocking(nullptr), options), WTFMove(completionHandler));
 }
 
 void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, const WebCore::Cookie& cookie, CompletionHandler<void(bool)>&& completionHandler) const
 {
-    RefPtr webFrame = document.frame() ? WebFrame::fromCoreFrame(*document.frame()) : nullptr;
-    if (!webFrame || !webFrame->page()) {
+    auto frame = document.frame();
+    if (!frame) {
         completionHandler(false);
         return;
     }
 
+    std::optional<WebCore::FrameIdentifier> frameID = std::nullopt;
+    std::optional<WebCore::PageIdentifier> pageID = std::nullopt;
+
     ApplyTrackingPrevention applyTrackingPreventionInNetworkProcess = ApplyTrackingPrevention::No;
 #if ENABLE(TRACKING_PREVENTION)
-    if (shouldBlockCookies(webFrame.get(), document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess)) {
-        completionHandler(false);
-        return;
+    if (frame->loader().client().isRemoteWorkerFrameLoaderClient()) {
```

Comment:
ditto, we should not special case remote workers. It should be a single code path.